File: 128_longest_consecutive_sequence.rb
# @param {Integer[]} nums
# @return {Integer}
def longest_consecutive(nums)
  return 0 if nums.length == 0
  
  max = 0
  
  present = Hash.new
  
  nums.each { |num| present[num] = true }
  
  nums.each do |num|
    count = 0
    
    unless present[num-1]
      curr_num = num
      while present[curr_num]
        count += 1
        curr_num += 1
      end
      
      if count > max
        max = count
      end
    end
  end

  max
end

File: 128_longest_consecutive_sequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  let counts = {};
  let result = 0;

  nums.forEach((num) => {
    counts[num] = true;
  });

  for (let key in counts) {
    key = parseInt(key);

    let subCount = 1;

    if (!counts[key - 1]) {
      while (counts[key + subCount]) {
        subCount++;
      }

      if (subCount + 1 > result) {
        result = subCount;
      }
    }
  }

  return result;
};